### DISCUSSION TITLE
Sonoff iFan04-L (and probably iFan03): slight improvement

### DISCUSSION THREAD

**Discussion by mruffalo** [RavenSystem/esp-homekit-devices#2056] [MEPLHAA Scripts Collection]
Hello-

First, thank you @RavenSystem for the *fantastic* firmware; I've installed it on four iFan04-L devices and it works perfectly; it's great to have native HomeKit control of a ceiling fan and light without pulling new wire through walls and ceilings.

I made a slight improvement to the iFan03/iFan04 script listed under [the Devices Database page](https://github.com/RavenSystem/esp-homekit-devices/wiki/Devices-Database), and wanted to share.

```python
{"c": {"io": [[[9, 12, 14, 15, 13], 2], [[0], 6]], "z": 12, "r": [{"n": 10, "s": 9600, "l": [8, 8]}], "l": 13, "b": [{"g": 0, "t": 5}]}, "a": [{"t": 65, "e": 3, "0": {"r": [{"g": 15}, {"g": 12}, {"g": 14}]}, "y0": [{"v": 1, "0": {"r": [{"g": 15, "v": 1, "i": 0.5}, {"g": 12}, {"g": 14, "i": 0.6}]}}, {"v": 2, "0": {"r": [{"g": 15, "v": 1, "i": 0.5}, {"g": 12, "i": 0.6}, {"g": 14, "i": 0.6}]}}, {"v": 3, "0": {"r": [{"g": 15, "v": 1}, {"g": 12}, {"g": 14, "v": 1}]}}], "es": [{"s": 1, "1": {"r": [{"g": 9}]}, "0": {"r": [{"g": 9, "v": 1}]}, "b": [{"g": 0}]}]}, {"t": 80, "h": 0, "n": 24, "pt": [["AA", 5]], "dt": [2, 0], "y0": [{"v": 260, "r": 1, "0": {"m": [[7002, 1]]}}, {"v": 517, "r": 1, "0": {"m": [[7002, 1]]}}, {"v": 6, "r": 1, "0": {"m": [[7001, 1], [-2]]}}, {"v": 263, "r": 1, "0": {"m": [[7001, 1], [-2, 1], [-2, 101]]}}, {"v": 520, "r": 1, "0": {"m": [[7001, 1], [-2, 2], [-2, 101]]}}, {"v": 777, "r": 1, "0": {"m": [[7001, 1], [-2, 3], [-2, 101]]}}, {"v": 1034, "r": 1, "0": {"m": [[7001, 1], [-1, 4]]}}, {"v": 265, "r": 1, "0": {"m": [[7004, 1]]}}, {"v": 266, "r": 1, "0": {"m": [[7003, 1]]}}]}, {"h": 0, "i": 0.1, "1": {"u": [{"v": "AA550104000005"}]}}, {"h": 0, "i": 0.1, "1": {"u": [{"v": "AA550101000002"}]}}, {"h": 0, "i": 0.1, "1": {"u": [{"v": "AA550107000008"}]}}, {"h": 0, "i": 0.1, "1": {"u": [{"v": "AA550106000007"}]}}]}
```

This adds the `"e"` option to set the total number of speed steps to 3, and adjusts both mappings:
* fan state → GPIO control (setting the motor speed), and
* RF remote → fan state

to use values 0 → 3 instead of 0 → 100. In my opinion this is a much better user experience than allowing any value from 0 → 100 with "hidden" thresholds corresponding to the three actual speeds implemented in the device's hardware.

I've tested this on the four iFan04-L devices I've installed, under firmware 12.4.4, and it works very well from both Apple Home and the RF remotes. I did have to reset the HomeKit ID, delete the devices from Home, and re-add them though -- otherwise the Home app still considered the devices to have 100 breakpoints, with 1%, 2%, and 3% corresponding to the three fan speeds 😅

**Comment by mruffalo** [comment]
Version with all whitespace removed:
```python
{"c":{"io":[[[9,12,14,15,13],2],[[0],6]],"z":12,"r":[{"n":10,"s":9600,"l":[8,8]}],"l":13,"b":[{"g":0,"t":5}]},"a":[{"t":65,"e":3,"0":{"r":[{"g":15},{"g":12},{"g":14}]},"y0":[{"v":1,"0":{"r":[{"g":15,"v":1,"i":0.5},{"g":12},{"g":14,"i":0.6}]}},{"v":2,"0":{"r":[{"g":15,"v":1,"i":0.5},{"g":12,"i":0.6},{"g":14,"i":0.6}]}},{"v":3,"0":{"r":[{"g":15,"v":1},{"g":12},{"g":14,"v":1}]}}],"es":[{"s":1,"1":{"r":[{"g":9}]},"0":{"r":[{"g":9,"v":1}]},"b":[{"g":0}]}]},{"t":80,"h":0,"n":24,"pt":[["AA",5]],"dt":[2,0],"y0":[{"v":260,"r":1,"0":{"m":[[7002,1]]}},{"v":517,"r":1,"0":{"m":[[7002,1]]}},{"v":6,"r":1,"0":{"m":[[7001,1],[-2]]}},{"v":263,"r":1,"0":{"m":[[7001,1],[-2,1],[-2,101]]}},{"v":520,"r":1,"0":{"m":[[7001,1],[-2,2],[-2,101]]}},{"v":777,"r":1,"0":{"m":[[7001,1],[-2,3],[-2,101]]}},{"v":1034,"r":1,"0":{"m":[[7001,1],[-1,4]]}},{"v":265,"r":1,"0":{"m":[[7004,1]]}},{"v":266,"r":1,"0":{"m":[[7003,1]]}}]},{"h":0,"i":0.1,"1":{"u":[{"v":"AA550104000005"}]}},{"h":0,"i":0.1,"1":{"u":[{"v":"AA550101000002"}]}},{"h":0,"i":0.1,"1":{"u":[{"v":"AA550107000008"}]}},{"h":0,"i":0.1,"1":{"u":[{"v":"AA550106000007"}]}}]}
```

**Comment by IsmaEC** [comment]
Hi, Do you have in RF Control 3 botton extra?
When I include RF Control, I have a remote control in HK but, Single press and Double press it is ok however Long Press not run. And I dont know why.
![image](https://github.com/RavenSystem/esp-homekit-devices/assets/137508011/bf0979ce-01ad-45f9-b00d-ee0f5f393cd0)

```json
{"c":{"n":"Ifan-Despacho”,”io":[[[9,12,14,15,13],2],[[0],6]],"z":12,"r":[{"n":10,"s":9600,"l":[8,8]}],"l":13,"b":[{"g":0,"t":5}]},"a":[{"0":{"r":[{"g":15},{"g":12},{"g":14}]},"t":65,"e":3,"y0":[{"0":{"r":[{"g":15,"v":1,"i":0.5},{"g":12},{"g":14,"i":0.6}]},"v":1},{"0":{"r":[{"g":15,"v":1,"i":0.5},{"g":12,"i":0.6},{"g":14,"i":0.6}]},"v":2},{"0":{"r":[{"g":15,"v":1},{"g":12},{"g":14,"v":1}]},"v":3}],"es":[{"0":{"r":[{"g":9,"v":1}]},"1":{"r":[{"g":9}]},"s":5,"b":[{"g":0}]}]},{"t":80,"h":0,"n":24,"pt":[["AA",5]],"dt":[2,0],"y0":[{"0":{"m":[[7002,1],[8,2]]},"v":260,"r":1},{"0":{"m":[[7002,1],[8,1]]},"v":517,"r":1},{"0":{"m":[[7001,1],[-2]]},"v":6,"r":1},{"0":{"m":[[7001,1],[-2,1],[-2,101]]},"v":263,"r":1},{"0":{"m":[[7001,1],[-2,2],[-2,101]]},"v":520,"r":1},{"0":{"m":[[7001,1],[-2,3],[-2,101]]},"v":777,"r":1},{"0":{"m":[[7001,1],[-1,4]]},"v":1034,"r":1},{"0":{"m":[[7004,1],[8,0]]},"v":265,"r":1},{"0":{"m":[[7003,1]]},"v":266,"r":1}]},{"1":{"u":[{"v":"AA550104000005"}]},"h":0,"i":0.1},{"1":{"u":[{"v":"AA550101000002"}]},"h":0,"i":0.1},{"1":{"u":[{"v":"AA550107000008"}]},"h":0,"i":0.1},{"1":{"u":[{"v":"AA550106000007"}]},"h":0,"i":0.1},{"t":3}]}
```

**Comment by mruffalo** [comment]
My only modifications to the script in the Devices Database page are:
* adding `"e": 3` to only allow three fan speeds, with internal values 0, 1, 2, 3
* Updating the mapping from internal values to GPIO settings to use those thresholds instead of 1, 33, and 67
* Adjusting the fan state values set by the RF remote to be 1, 2, 3 instead of 30, 60, 100

Everything else is unchanged from the previous script, including whatever behavior those buttons have/had.

I actually don't know what those buttons do, except for (I believe) the bottom left one which you marked as "doble press", which disassociates the remote from the RF controller in the unit, necessitating pairing it again after killing power to the iFan04-L. (This is especially painful for the two units I have on our second floor, which are on the same breaker.)

**Reply by vypergts** [reply]
@mruffalo I used your script but for the life of me I cannot get the RF remote to pair. Is there a secret? I also have to flip the breaker a in the basement for the fan on the second floor.

**Comment by edValenciaMX** [comment]
Hi @mruffalo,

I'm new to the HAA, I not even install any device yet but I'm a Apple Home user and I want to flash my iFan-04L, but I did not find any tutorial or guide on how to do it, do you know any? I was very confuse with the tasmota an ESP firmware and I was afraid to brick the device but seems posible to do it. I have 2 devices working with AMZ Alexa, but I hate that it does not have multi home for the same account, the worst. And google Home its not very useful in my country.

Thanks, I would really appreciate your help.

**Reply by IsmaEC** [reply]
I have a tutorial but is writing in Spanish, the json is not well because in V12 it change.
You can read it here:

https://drive.google.com/file/d/1RtMMUzmv-oAZBv56i-u0-LQCCayNfNCR/view?usp=drivesdk


If you have any question in Discord I'm @IsmaelEC

**Reply by edValenciaMX** [reply]
Thank you!

I've read it, luckily Spanish its my primary language, I might need a serial TTL to try it but seems cheap to buy.

**Comment by ururk** [comment]
Hi @mruffalo,

I'm getting ready to flash an iFan04-L, and was comparing your config to the one posted on the devices database. The "official" configs use arrays where your config uses objects (ie, `"r":[[9,1]]` vs `"r":[{ "g": 9,"v": 1}]`) - was wondering if you know if this was a change with how devices are now configured, or if both are supported.

Also - not having used a fan with HomeKit, are you basically saying that there will always be 100 steps for a fan control in HomeKit - and either the setpoints will be at `0%, 33%, 67% 100%` with the default config, versus `0%, 1%, 2%, 3%` with your config? Or does HomeKit have steps (ie, low / medium / high) that one can map to these %'s?
